lib/rules/jsx-classname-namespace.js

```diff
@@ -0,0 +1,156 @@
+ * @author Automattic
+ * @copyright 2016 Automattic. All rights reserved.
+ * See LICENSE.md file in root directory for full license.
+ */
+'use strict';
+
+//------------------------------------------------------------------------------
+// Requirements
+//------------------------------------------------------------------------------
+
+var path = require( 'path' );
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+var rule = module.exports = function( context ) {
+	function isModuleExportNode( node ) {
+		return (
+			'ExpressionStatement' === node.type &&
+			'AssignmentExpression' === node.expression.type &&
+			'MemberExpression' === node.expression.left.type &&
+			'module' === node.expression.left.object.name &&
+			'exports' === node.expression.left.property.name
+		);
+	}
+
+	function isIdentifierInDescendents( node, name ) {
+		switch ( node.type ) {
+			case 'Identifier':
+				return node.name === name;
+
+			case 'CallExpression':
+				return node.arguments.some( function( argNode ) {
+					return isIdentifierInDescendents( argNode, name );
+				} );
+		}
+
+		return false;
+	}
+
+	function isRootRenderedElement( node ) {
+		var parent = node.parent.parent.parent,
+			functionExpression, functionName;
+
+		// If wrapped in a JSX element, the node is a child
+		if ( 'JSXElement' === parent.type ) {
+			return false;
+		}
+
+		do {
+			if ( -1 !== [ 'FunctionExpression', 'FunctionDeclaration', 'ArrowFunctionExpression' ].indexOf( parent.type ) ) {
+				functionExpression = parent;
+
+				// If inside function expression, check that the name of the
+				// property is "render" (React.createClass or Component class)
+				if ( 'FunctionExpression' === parent.type &&
+						'Property' === parent.parent.type &&
+						'init' === parent.parent.kind &&
+						'render' === parent.parent.key.name ) {
+					return true;
+				}
+
+				// If wrapped in function declaration, check the declaration
+				// is part of a default export (stateless function component)
+				while ( parent.parent ) {
+					parent = parent.parent;
+					if ( 'ExportDefaultDeclaration' === parent.type ||
+							isModuleExportNode( parent ) ) {
+						return true;
+					}
+				}
+
+				switch ( functionExpression.type ) {
+					case 'FunctionDeclaration':
+						if ( functionExpression.id ) {
+							functionName = functionExpression.id.name;
+						}
+						break;
+
+					case 'ArrowFunctionExpression':
+						if ( 'VariableDeclarator' === functionExpression.parent.type ) {
+							functionName = functionExpression.parent.id.name;
+						}
+						break;
+				}
+
+				// If we've exhausted parent options, check to see whether
+				// program's exports refer to this expression
+				return functionName && 'Program' === parent.type && parent.body.some( function( programNode ) {
```

Comment:
It's a bit pedantic, but when there's no functionName `isRootRenderedElement()` returns `undefined`, but it returns either `true` or `false` in every other case.  `functionName !== undefined` here would make for return values more consistent.